api/client/proxy.go

```diff
@@ -28,23 +29,34 @@ import (
 )
 
 // DialProxy creates a connection to a server via an HTTP Proxy.
-func DialProxy(ctx context.Context, proxyAddr, addr string) (net.Conn, error) {
-	return DialProxyWithDialer(ctx, proxyAddr, addr, &net.Dialer{})
+func DialProxy(ctx context.Context, proxyURL *url.URL, addr string) (net.Conn, error) {
+	return DialProxyWithDialer(ctx, proxyURL, addr, &net.Dialer{})
 }
 
 // DialProxyWithDialer creates a connection to a server via an HTTP Proxy using a specified dialer.
-func DialProxyWithDialer(ctx context.Context, proxyAddr, addr string, dialer ContextDialer) (net.Conn, error) {
-	conn, err := dialer.DialContext(ctx, "tcp", proxyAddr)
+func DialProxyWithDialer(ctx context.Context, proxyURL *url.URL, addr string, dialer ContextDialer) (net.Conn, error) {
+	conn, err := dialer.DialContext(ctx, "tcp", proxyURL.Host)
 	if err != nil {
-		log.Warnf("Unable to dial to proxy: %v: %v.", proxyAddr, err)
+		log.Warnf("Unable to dial to proxy: %v: %v.", proxyURL.Host, err)
 		return nil, trace.ConvertSystemError(err)
 	}
 
+	header := make(http.Header)
+	if proxyURL != nil && proxyURL.User != nil {
+		// dont use User.String() because it performs url encoding (rfc 1738),
+		// which we don't want in our header
+		creds := proxyURL.User.Username()
+		if password, hasPass := proxyURL.User.Password(); hasPass {
+			creds += ":" + password
```

Comment:
Does there is any difference in handling user without password case 
Namely  we will set basicAuth to "user" where GO https://github.com/golang/go/blob/690ac4071fa3e07113bf371c9e74394ab54d6749/src/net/http/transport.go#L849
in that case adds the ":" token and the end of user (`user:`)    ?

Reply:
Hmm, I went back and forth on this actually. I saw that in User.String() it only adds the ":" if the password is set, so I went with that, but I see they always add it in the code you linked for proxy auth header. Looking at the history for that code, it looks like they used to conditionally add the ":" but changed it to "fix basic authentication", so that must be the right way.